Add forceInclude option for `:where` selectors

batbattur opened the pull request:
## Summary
For some reason, penthouse doesn't think that `:where` selectors are not part of the critical css. 
Let's add it to the config so that it includes the `:where` selectors in the generated critical css.

Edit: The new regex will match selectors like the following: `:where()`, `:has()`, `:not()`, `:host()`, `:root()`.

See https://github.com/iFixit/ifixit/pull/47766#issuecomment-1544460438 for more info.

## QA notes
We can qa this on `iFixit/ifixit` side by running the `critical.css` workflow (with this branch checked out). 

cc: @ianrohde @mlahargou

batbattur committed:
Add forceInclude option for :where selectors

For some reason, penthouse doesn't think that :where selectors are not
part of the critical css. Let's add it to the config so that it includes
the :where selectors in the generated critical css.

See https://github.com/iFixit/ifixit/pull/47766#issuecomment-1544460438 for more info.

batbattur commented:
dev_block 🚛 on https://github.com/iFixit/ifixit/pull/47766#issuecomment-1544475631

sctice-ifixit commented on critical-css.js:
```diff
@@ -46,6 +46,7 @@ function findCriticalCss(cssString, url) {
     height: 2160,
     blockJSRequests: false,
     forceExclude: excludeArr,
+    forceInclude: [/where/i],
```
As I understand it, this is going to force including any CSS selector that matches. I think we want to be as specific as we can be here because I don't think it's necessarily the case that every use of `:where()` is critical CSS. Beyond that, I think `/where/` is too broad as that would match selectors like `.where-empty`.

batbattur replied:
>  I think we want to be as specific as we can be here because I don't think it's necessarily the case that every use of :where() is critical CSS.

I agree. In the future, if we want to exclude some of the `:where()` uses, then we can exclude them by specifically adding to this array in our [exclude array in critical.css.json config](https://github.com/iFixit/ifixit/blob/bdca6bd6862a54066c52771d94c023e9109d2db2/Cart/UI/templates/css/FeaturedProduct/MantaDriverKit/critical.css.json#L5) like the following: `"excludes": [":where(.someSpecificSelector)"]`


> Beyond that, I think /where/ is too broad as that would match selectors like .where-empty.

I modified the regex to match other css features as requested in https://github.com/iFixit/ifixit/pull/47766#issuecomment-1544475631. Updated in https://github.com/iFixit/css-gather/pull/11/commits/80656ab3c9a33edfdca9ffa7b77dbeebd4eb3746.

The new regex will match selectors like the following: `:where()`, `:has()`, `:not()`, `:host()`, `:root()`.

Verified that it will not match selectors like the following: `li(:last-child)`, `.green:not(:disabled)`, and `.card-hover:hover`.

sctice-ifixit replied:
I'm still not convinced that this is an appropriate solution. The point of the tool is that _it_ identifies which CSS ultimately affects above-the-fold content. We're unconditionally insisting that any of these selectors _always_ affects above-the-fold content, which is surely false. While that may be a solution to the problem we were seeing, I think at best it's papering over some deeper issue, either with penthouse or css-gather or our styles in general.

If this fixes the problem we were seeing (CLS), then as a temporary measure I think we could add more-specific `forceInclude` selectors that, when omitted, cause the CLS that we're currently seeing. But as-is, this pull has the potential to add a bunch of CSS that isn't critical to our critical CSS, and that means sending more bytes than we need to on every page load because we send that CSS directly in the HTML.

probablyian replied:
It's not optimal, but our options are limited. Your points are all valid, but without these changes, the resets are removed entirely. That would cause major CLS on every page.

I looked at the [penthouse repo](https://github.com/pocketjoso/penthouse) and the last closed issue was about a year ago. I doubt the maintainer is still active. Should we spin up an issue to address this more thoroughly? I'd be happy to contribute here if it would help.

mlahargou replied:
They are on a super old version of puppeteer. Might be related.

https://github.com/pocketjoso/penthouse/pull/347

batbattur replied:
I created an issue with minimal repro steps in their repo: https://github.com/pocketjoso/penthouse/issues/356.
Temporarily, we could use `forceInclude` until that issue is fixed on the penthouse's side, or possibly revert https://github.com/iFixit/ifixit/pull/47597.

probablyian replied:
I don't see any reason to change or revert based on the output here: https://github.com/iFixit/ifixit/pull/47846/files

batbattur replied:
Yes, the new pull seems to picking up the selectors correctly now.

sctice-ifixit replied:
> without these changes, the resets are removed entirely. That would cause major CLS on every page.

I'm not suggesting that we don't use `forceInclude` (although I think that would be the best outcome). We can specify exactly the selectors that we need instead of any selector matching a broad regex. The changes as they are apply to far more than the current reset rules that we care about; they apply to _all_ matching selectors, present and future. Even if that turns out to be the same set (right now) as targeted by a more specific set of selectors, it has the potential to match any future CSS that happens to use `:…(`. I don't see any harm in being explicit until we run into this problem repeatedly and find that we're wasting time going and adding more specific selectors.

It's okay with me if you want to ignore this concern in order to move quickly, but I think we should probably come back and address it sooner rather than later. We're creating a small liability in that every pull that introduces CSS to pages where we use critical CSS now has the potential to ship users CSS they don't need on every page load in a way that will likely be invisible to us.

> They are on a super old version of puppeteer. Might be related.

That seems like a good hypothesis. If it's old enough that the underlying browser doesn't recognize the selectors and match them against the DOM, the tool probably wouldn't be able to see the selectors as used.

batbattur replied:
> it has the potential to match any future CSS that happens to use :…(. 

Yes, I definitely agree with this. 

<hr>

We could use a regex to match `:where(` like `forceInclude: [/^:where\(/]` 

or be very explicit and use the string format like the following to match them exactly from the [assets](https://assets.cdn.ifixit.com/Assets/Guide/ifixit/guide-all-baQyntVbn71J5gJenC0L9w.css) and from this [pull](https://github.com/iFixit/ifixit/pull/47597/files): 
```
    forceInclude: [
      ":where(*,::before,::after)",
      ":where(*)",
      ":where(html)",
      ":where(body)",
      ":where(button,input,select,textarea)",
      ":where(button)",
      ":where(input)"
    ],
```

Any thoughts on which one we prefer?

sctice-ifixit replied:
I think starting with the literal strings is prudent unless we're confident that we'll be adding more exceptions on a regular basis. Since we're addressing the symptom rather than the cause, I think our intervention should be as specific as we can get away with while balancing against things like time, effort, and so on.

probablyian replied:
I'm ok with being explicit, but we are going to have to be careful moving forward if the reset ever changes. I was hoping to avoid manual work (hence the regex), but I don't see a median in this situation. Perhaps we can put a note in the automated PR description to check if there have been any reset changes since the previous week's run? It rarely changes, so that may be overkill.

sctice-ifixit replied:
> I was hoping to avoid manual work (hence the regex), but I don't see a median in this situation.

Totally fair. I'd just like us to prove that we're going to need more manual work before we try to solve for that. I'm hoping that we figure out a way to address the root cause before that becomes a problem. I think that [@mlahargou's theory about being on an old version of puppeteer](https://github.com/iFixit/css-gather/pull/11#discussion_r1191662842) is a promising avenue to pursue. Ideally that gets fixed upstream, but perhaps we just need to fork and apply the pull linked in that comment.

batbattur replied:
I just verified that https://github.com/pocketjoso/penthouse/pull/347 fixes the problem.

Manually updated the penthouse in `package.json` to point at the fork (with the updated puppeteer version) from the comment and re-ran the css-gather. The generated critical-css now includes the `:where` selectors.
```diff
         "cssnano-preset-advanced": "^5.0.0",
-        "penthouse": "^2.3.2",
+        "penthouse": "feedanal/penthouse#master",
         "postcss": "^8.2.10",
```

In that case, should we just close this pull and make a fork for penthouse? Then when https://github.com/pocketjoso/penthouse/pull/347 gets fixed we can drop the fork. That seems like a better idea.

batbattur replied:
Going to close this pull, create a fork for `penthouse` with updated puppeteer, and use that fork for `css-gather`. 👍

sctice-ifixit replied:
Thanks, @batbattur!

batbattur committed:
Update the forceInclude regex to be specific

This new regex will match selectors like the following:
:where(), :has(), :not(), :host(), :root().

Verified that it will not match selectors like the following:
li(:last-child), .green:not(:disabled), .card-hover:hover.

batbattur commented:
Steps used to test:
1. Create a test branch to run the `critical-css.yml` from this branch: https://github.com/iFixit/ifixit/compare/testing-critical-css-forceInclude
2. Ran `gh workflow run critical-css.yml --ref=testing-critical-css-forceInclude` from cominor. Then it created https://github.com/iFixit/ifixit/actions/runs/4951784687 run.
3. Finally, it created a pull request with the changes from this pull: https://github.com/iFixit/ifixit/pull/47846/files
4. Made sure the `:where(...)` selectors are included in the generated critical-css files.

QA 🚛

mlahargou reviewed (approved):
CR 📱
